### DISCUSSION TITLE
@Unique decorator

### DISCUSSION THREAD

**Discussion by Lishenga** [mikro-orm/mikro-orm#1735] [Q&A]
Hi, i need some clarification, when i use the @Unique at the entity level not at the column level, do i mean that column _name_ and _email_ should not have the same values under 1 record in the table or that 2 or more records under the specified table, the columns _name_ and _email_ should not have the same values. Let me try to elaborate the two instances

- For the columns _name_ and _email_, a record in the DB for the specified table **shouldn't** have values for column _name_ and _email_ like below
```json
"name": "brian@gmail.com",
"email": "brian@gmail.com"
```
Hence the column values for _name_ and _email_ should be unique or do you mean the second instance or scenario, which is

- For the columns _name_ and _email_, two or more records for the specified table **shouldn't** be same or equal for the two columns but the two columns can be equal or same under the same record for the specified table. For example
```json
"name": "brian@gmail.com",
"email": "brian@gmail.com"
```
```json
"name": "brian@gmail.com",
"email": "leon@gmail.com"
```
The above two records have column _name_ having the same values. Does the @Unique decorator limit similarity on the values of the column _name_ on two or more records in the table or it limits on the similarity of the values on **only** the two columns _name_ and _email_ for a **single** record under the table ?

The table that contains the records is 

```ts

@Entity({ tableName: 'item', customRepository: () => ItemRepository })
@Index({ name: 'emailName', properties: ['name', 'email'] }) // simple index, with custom name
@Unique({ properties: ['name', 'email'] })
export class Item extends Thing {

  [EntityRepositoryType]?: ItemRepository;

  @Field({ defaultValue: "OVES", description: "Default ItemID is 'OVES'" })
  @Property()
  public name: string;
 
 
  @Field()
  @Property()
  public email: string;

}

```

**Comment by B4nan** [comment]
Its composite index, so its always about the combination of the fields. 

https://stackoverflow.com/questions/795031/how-do-composite-indexes-work

**Reply by Lishenga** [reply]
Thank you for the answer, so is there a decorator in mikroorm that makes the values in a column be different across different records in a database? Whereby if a new record is being created with a value that exists in another record on the "unique" column, mikroorm will through an error stating the provided value already exists in another record on the 'unique' column under the same table. Something like two users can't exist in a system with same email address, the two users must have unique email addresses.

**Reply by B4nan** [reply]
Don't be too concerned with the ORM, and search how things work in the underlying database. ORM is just about mapping those things. It's not the ORM who checks this, its the database itself.

If you want to constraint single property, then don't use composite index. Simple as that.

**Reply by Lishenga** [reply]
Okay thank you any idea of how i could constraint a single property?

**Reply by B4nan** [reply]
```ts
@Unique()
@Property()
email: string;
```

or 

```ts
@Property({ unique: true })
email: string;
```

or at the entity level

```ts
@Unique({ properties: 'email' }) // or @Unique({ properties: ['email'] })
@Entity()
export class User {

  @Property()
  email: string;

}
```

**Reply by Lishenga** [reply]
What is the difference then between @Unique and @Index. Because from your documentation I see you stated that if i want to use unique indexes i use at @Unique but from your example above it seems that i can also use @Unique to constrain a column. In line with how does @Index give me the capability of creating unique indexes from two columns. Lastly, does @Unique in constrain a specific column in a MongoDB? 

Sorry for many questions, am a little bit confused. I would highly appreciate your assistance.

**Reply by B4nan** [reply]
Again, learn how things work in the database you are using, don't be too concerned with the ORM (in other words, this is out of scope of the ORM docs). 

The naming should be quite obvious - you want something unique, use `@Unique` decorator. You want to index something (make reads/queries faster), use `@Index` decorator.
